scanner/engine/master.cpp

```diff
@@ -215,218 +226,262 @@ class MasterImpl final : public proto::Master::Service {
+    job_descriptor.mutable_tasks()->CopyFrom(tasks);
+
+    validate_task_set(meta, job_params->task_set(), job_result);
+    if (!job_result->success()) {
+      // No database changes made at this point, so just return
+      return grpc::Status::OK;
+    }
+
+    // Add job name into database metadata so we can look up what jobs have
+    // been ran
+    i32 job_id = meta.add_job(job_params->job_name());
+    job_descriptor.set_id(job_id);
+    job_descriptor.set_name(job_params->job_name());
+
+    // Read all table metadata
+    for (const std::string &table_name : meta.table_names()) {
+      std::string table_path =
+          TableMetadata::descriptor_path(meta.get_table_id(table_name));
+      table_metas_[table_name] = read_table_metadata(storage_, table_path);
+    }
+
+    for (auto &task : job_params->task_set().tasks()) {
+      i32 table_id = meta.add_table(task.output_table_name());
+      proto::TableDescriptor table_desc;
+      table_desc.set_id(table_id);
+      table_desc.set_name(task.output_table_name());
+      table_desc.set_timestamp(std::chrono::duration_cast<std::chrono::seconds>(
+                                   now().time_since_epoch())
+                                   .count());
+      // Set columns equal to the last op's output columns
       for (size_t i = 0; i < output_columns.size(); ++i) {
-        auto &col_name = output_columns[i];
-        Column *col = job_descriptor.add_columns();
+        Column *col = table_desc.add_columns();
         col->set_id(i);
-        col->set_name(col_name);
+        col->set_name(output_columns[i]);
         col->set_type(ColumnType::Other);
       }
+      table_metas_[task.output_table_name()] = TableMetadata(table_desc);
+      std::vector<i64> end_rows;
+      Result result = get_task_end_rows(table_metas_, task, end_rows);
+      if (!result.success()) {
+        *job_result = result;
+        break;
+      }
+      for (i64 r : end_rows) {
+        table_desc.add_end_rows(r);
+      }
+      table_desc.set_job_id(job_id);
 
-      DatabaseMetadata meta =
-          read_database_metadata(storage_, DatabaseMetadata::descriptor_path());
-
-      auto &tasks = job_params->task_set().tasks();
-      job_descriptor.mutable_tasks()->CopyFrom(tasks);
+      write_table_metadata(storage_, TableMetadata(table_desc));
+      table_metas_[task.output_table_name()] = TableMetadata(table_desc);
+    }
+    if (!job_result->success()) {
+      // No database changes made at this point, so just return
+      return grpc::Status::OK;
+    }
 
-      validate_task_set(meta, job_params->task_set(), job_result);
-      if (!job_result->success()) {
-        // No database changes made at this point, so just return
-        return grpc::Status::OK;
-      }
+    // Write out database metadata so that workers can read it
+    write_job_metadata(storage_, JobMetadata(job_descriptor));
+
+    // Setup initial task sampler
+    task_result_.set_success(true);
```

Comment:
Why is it necessary to set up a task sampler here? Couldn't you set `samples_left_ = 0` and then the logic in `NewWork` will take care of it?